compute_tools/src/monitor.rs

```diff
@@ -0,0 +1,107 @@
+use std::sync::{Arc, RwLock};
+use std::{thread, time};
+
+use anyhow::Result;
+use chrono::{DateTime, Utc};
+use log::{debug, info};
+use postgres::{Client, NoTls};
+
+use crate::zenith::ComputeState;
+
+// Spin in a loop and figure out the last activity time in the Postgres.
+// Then update it in the shared state. This function never errors out.
+// XXX: the only expected panic is at `RwLock` unwrap().
+fn watch_compute_activity(state: &Arc<RwLock<ComputeState>>) {
+    // Suppose that `connstr` doesn't change
+    let connstr = state.read().unwrap().connstr.clone();
+    // Define `client` outside of the loop to reuse existing connection if it's active.
+    let mut client = Client::connect(&connstr, NoTls);
+
+    info!("watching Postgres activity at {}", connstr);
+
+    loop {
+        // Should be before write lock to allow others to read while we sleep.
+        thread::sleep(time::Duration::from_millis(1000));
+
+        match &mut client {
+            Ok(cli) => {
+                if cli.is_closed() {
+                    info!("connection to postgres closed, trying to reconnect");
+
+                    // Connection is closed, reconnect and try again.
+                    client = Client::connect(&connstr, NoTls);
+                    continue;
```

Comment:
Yeah, makes sense, thanks

Reply:
On the second thought...
> Connection is established so looks like we can proceed with new connection
Do we? `Client::connect` can easily return error / closed connection. Later query will error out if it's closed, but I thought that it'd be just more clean to check whether we established a connection or not. And the easiest way to do it here without code duplication it just to continue

Reply:
> And the easiest way to do it here without code duplication it just to continue

If there is a function named something like `ensure_connected` which checks current connection and makes a new one if needed no code duplication is involved and all intentions are clean. Sometimes it is handy to have loop and it's iteration in different functions so you can use early returns and error propagation with `?`. This way we can eliminate extra delay between the error occurrence and a log message. (if these are far apart in logs this may complicate investigation when this is correlated from logs in different service)